### ISSUE TITLE
Results cannot improve by rounding (.next): Federated TF 

### ISSUE THREAD

**Issue by AbbasiAYE** [tensorflow/federated#493]
Hello

I'm trying to compare federated learning TF with keras fitting, as we see below. However, the results of TFF does not improve by rounding. Also, when I set weight a local model based on the TFF model, I get worse results than normally training the model on keras. The Loss results are mentioned after the code.

Thanks for the help.

Regards

# Code:
tf.compat.v1.enable_v2_behavior()

TobeRemoved_Array = ['feature1','feature2','feature3','feature4','feature5','feature6','feature7', 'is_training']

train_perc = 0.8
Norm_Input  = True
Norm_Output = False
Input_str  = ['feature1', 'feature2']
if Norm_Output:
    Output_str = ['feature3']
else: 
    Output_str = ['feature4']

Final_Tag = 'Tag3' # here just decide which tagging method do you want to use
Num_Clients = 10 # cannot be less than one
Num_Cons_Sample_Client = 5 # cannot be less than one

# Load simulation data.
##############################################
dir_name = 'pickle-data/'
file_name = 'logs_april_2019.pickle'
files = os.listdir('pickle-data/')
dataframe = Import_Pickle.Import_v1(dir_name,file_name,False) # choose False to use 2019 data
# Just to reduce the processing
ave = dataframe.core.min() + 1000000000
df2 = dataframe[dataframe.core < ave]
df = Import_Pickle.PreProcessing_v2019(df2,Norm_Input)
train_df,test_df,X_traindf,X_testdf,Y_traindf,Y_testdf,XY_traindf,XY_testdf = Import_Pickle.Splitting_Train_Test(df,train_perc,Norm_Output,TobeRemoved_Array)
########## splitting for clients ############  
def Tag_per_day(train_df_loc,TagNum):
    train_df_loc['log2'] =  train_df_loc['log'].apply(lambda x: x.replace("_",""))
    tag_Index = train_df_loc.log2.apply(lambda x: x.index("201"))
    tag_Index2 = tag_Index.values[1]
    tag_date =train_df_loc.log2.apply(lambda x: x[tag_Index2:tag_Index2+8])
    train_df_loc.loc[:,'Tag'+str(TagNum)] = pd.Series(tag_date.to_list(),index=train_df.index)  # to be fixed
    return train_df_loc

# Introduce time as input
X_traindf['feature1'] = train_df['feature1']
# introduce first tag per day
TagNum=1
train_df = Tag_per_day(train_df,TagNum)
#examples on groupby
Unq_tag1_grps = list(train_df.groupby(train_df.Tag1).groups.keys())
train_df.groupby(train_df.Tag1).first()
train_df.groupby(train_df.Tag1)['feature1'].count()
X_traindf['Tag'+str(TagNum)] =  train_df['Tag'+str(TagNum)]
#############################
# introduce epoch as tag
#############################
TagNum=2
train_df['Tag'+str(TagNum)] = train_df.epoch
X_traindf['Tag'+str(TagNum)] =  train_df['Tag'+str(TagNum)]
#############################
# introduce core as tag
#############################
TagNum=3
train_df['Tag'+str(TagNum)] = train_df.core
X_traindf['Tag'+str(TagNum)] =  train_df['Tag'+str(TagNum)]
#############################
# introduce day as tag per client
#############################
TagNum = 4
RepNum = np.ceil(train_df.shape[0]/(Num_Cons_Sample_Client*Num_Clients))
Part_Tag_Array=[]
for i in np.arange(Num_Clients):
    Part_Tag_Tmp = list(map(lambda _: i+1,range(Num_Cons_Sample_Client)))
    Part_Tag_Array.extend(Part_Tag_Tmp)

Full_Tag_Array2 = Part_Tag_Array * int(RepNum)
extra_tags = np.abs(len(Full_Tag_Array2) - train_df.shape[0])
Full_Tag_Array = Full_Tag_Array2[:-extra_tags]

train_df.loc[:,'Tag'+str(TagNum)] = pd.Series(Full_Tag_Array,index=train_df.index)
X_traindf.loc[:,'Tag'+str(TagNum)] = train_df['Tag'+str(TagNum)]
#############################
# END day as tag per client
#############################
######### Introduce gpsTime and Tag to the input
Input_str.extend(['feature1',Final_Tag])

###### Adding StandardSalarization:
scaler = StandardScaler()
removed_column = Input_str.pop()
X_train_ScaledTmp = scaler.fit_transform(X_traindf[Input_str],Y_traindf[Output_str])
# Adding Int tag per client without scalarization
X_train_Scaled = np.c_[X_train_ScaledTmp, train_df[removed_column].values.reshape(train_df.shape[0],1)]

# All In/Out data Numpy
Act_Inputs_Int_Tag  = X_train_Scaled
Act_Outputs_Int = Y_traindf[Output_str].values
# Remove Tags
Act_Inputs_Int = np.delete(Act_Inputs_Int_Tag,-1,axis=1) 

# prepare In/Out per Client
All_Act_Inputs_Int_Tag  = [Act_Inputs_Int_Tag[np.where(Act_Inputs_Int_Tag[:,-1]== x)] for x in np.arange(1,Num_Clients+1)]
All_Act_Outputs_Int = [Act_Outputs_Int[np.where(Act_Inputs_Int_Tag[:,-1]== x)] for x in np.arange(1,Num_Clients+1)]
# Remove Tags
All_Act_Inputs_Int = [np.delete(All_Act_Inputs_Int_Tag[x],-1,axis=1)  for x in np.arange(0,Num_Clients) ]

# a need conversion to float32
Act_Inputs = np.float32(Act_Inputs_Int)
Act_Outputs = np.float32(Act_Outputs_Int)
# convert dataset to client based dataset
All_Act_Inputs = [np.float32(All_Act_Inputs_Int[x]) for x in np.arange(0,Num_Clients)]
All_Act_Outputs = [np.float32(All_Act_Outputs_Int[x]) for x in np.arange(0,Num_Clients)]
# convert to OrderedDict
new_batch = collections.OrderedDict([('In', Act_Inputs),('Out', Act_Outputs)])
All_new_batch = [collections.OrderedDict([('In', All_Act_Inputs[x]),('Out', All_Act_Outputs[x])]) for x in np.arange(0,Num_Clients)]
# Convert to tensor
dataset_input = tf.data.Dataset.from_tensor_slices(new_batch)#,,maxval=100, dtype=tf.float32)
# All_new_batch has different item per In / Out
All_dataset_input = [tf.data.Dataset.from_tensor_slices(All_new_batch[x]) for x in np.arange(0,Num_Clients)]
# Select among the datasets
Used_dataset= dataset_input
All_Used_dataset= All_dataset_input

#with eager_mode():
learning_rate = 1
Val_Train_Split = 0.8
SNN_epoch = 50
SNN_batch_size = 1100
shuffle_buffer = 200

def preprocess(new_dataset):
    #return Used_dataset.repeat(2).batch(2)
    def map_fn(elem):
        return collections.OrderedDict([('x', tf.reshape(elem['In'], [-1])),('y', tf.reshape(elem['Out'],[1]))])
     
    DS2= new_dataset.map(map_fn)
    return DS2.repeat(SNN_epoch).batch(SNN_batch_size)

train_data = [preprocess(Used_dataset)]

#######changes###############33
def make_federated_data(client_data, client_ids):
    return [preprocess(client_data[x]) for x in client_ids]

federated_train_data =  make_federated_data(All_Used_dataset, np.arange(0,Num_Clients))

sample_batch = tf.contrib.framework.nest.map_structure(lambda x: x.numpy(), next(iter(train_data[0])))
    
########## END Changes ############            

def create_SK_model():
    modelF = tf.keras.Sequential([tf.keras.layers.Flatten(input_shape=(Act_Inputs.shape[1],)),
                                  tf.keras.layers.Dense(10, activation=tf.nn.relu, kernel_initializer='zeros'),
                                  tf.keras.layers.Dense(100, activation=tf.nn.relu, kernel_initializer='zeros'),
                                  tf.keras.layers.Dropout(0.5),
                                  tf.keras.layers.Dense(100, activation=tf.nn.relu, kernel_initializer='zeros'),
                                  tf.keras.layers.Dropout(0.5),
                                  tf.keras.layers.Dense(1, activation=tf.nn.relu, kernel_initializer='zeros'),
                                  ])
    return modelF

def loss_fn(y_true, y_pred):
    return tf.reduce_mean(tf.keras.losses.MSE(y_true, y_pred))

def model_fn():
    return tff.learning.from_keras_model(create_SK_model(), 
                                         sample_batch, loss=loss_fn, 
                                         optimizer=gradient_descent.SGD(0.1))

YTrain = Act_Outputs #np.random.rand(50,1)
XTrain = Act_Inputs  #np.random.rand(50,100)
# locally compile the model
Local_model = create_SK_model()
Local_model.compile(loss=loss_fn,optimizer=gradient_descent.SGD(learning_rate))

# training/fitting with TF federated learning
trainer_Itr_Process = tff.learning.build_federated_averaging_process(model_fn,server_optimizer_fn=(lambda : gradient_descent.SGD(learning_rate=1.0)),client_weight_fn=None)
FLstate = trainer_Itr_Process.initialize()
FLlosses  = []
# Track loss of different ...... of federated iteration
for round_num in range(2, 5): 
    """
    The second of the pair of federated computations, next, represents a single round of Federated Averaging, which consists of pushing the server state (including the model parameters) to the clients, on-device training on their local data, collecting and averaging model updates, and producing a new updated model at the server.
    """ 
    FLstate, FLoutputs = trainer_Itr_Process.next(FLstate, federated_train_data)
    print('round {:2d}, metrics={}'.format(round_num, FLoutputs.loss))
    # Track the loss.
    FLlosses.append(FLoutputs.loss)

# fitting without federated learning
trained_local_Model = Local_model.fit(XTrain,YTrain, validation_split=Val_Train_Split, epochs=SNN_epoch, batch_size=SNN_batch_size) #tbuc
# Loss of local model
local_Loss = trained_local_Model.history['loss'] # tbuc
print("Local Model Loss: "+str(local_Loss))
# Copy local model for comparison purposes
Local_model_Fed = Local_model
# Setting federated weights on copied Object of local model
Local_model_Fed.set_weights(tff.learning.keras_weights_from_tff_weights(FLstate.model))
# Evaluate loss of the copied federated weights on local model
Fed_predicted = Local_model_Fed.predict(XTrain)
Fed_eval = Local_model_Fed.evaluate(XTrain,YTrain)

print("Local Model with Federated Weights Loss: "+str(Fed_eval ))
print("New: " + str(Local_model_Fed.loss))
print('End')

# Output:
round  2, metrics=6838.99658203125
round  3, metrics=6838.99658203125
round  4, metrics=6838.99658203125
round  5, metrics=6838.99658203125
round  6, metrics=6838.99658203125
Train on 733 samples, validate on 2936 samples
Epoch 1/50

733/733 [==============================] - 0s 14us/sample - loss: 6449.8022 - val_loss: 6936.1622
Epoch 4/50
.........................................................
733/733 [==============================] - 0s 18us/sample - loss: 6449.8022 - val_loss: 6936.1622

Local Model Loss: [6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375]

  32/3669 [..............................] - ETA: 0s - loss: 10104.2188
1632/3669 [============>.................] - ETA: 0s - loss: 6897.5380 
3232/3669 [=========================>....] - ETA: 0s - loss: 6860.1061
3669/3669 [==============================] - 0s 31us/sample - loss: 6838.9962
Local Model with Federated Weights Loss: 6838.996183714057

**Comment by ZacharyGarrett** [comment]
It appears this model is not training even without federation, I believe the following lines show the loss during non-federated training,:

```
local_Loss = trained_local_Model.history['loss'] # tbuc
print("Local Model Loss: "+str(local_Loss))
```

The loss does not appear to be decreasing:

```
Local Model Loss: [6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375, 6449.80224609375]
```

TFF only extends local training to the federated setting; we would be very surprised if a model that will not train locally was trainable in the federated setting. Closing this as not a TFF issue.

I'd recommend first getting local training to work. StackOverflow with the [tensorflow] and [keras] tags can be good resources. Once local training is working, try adding back TFF for federated optimization.
